File: connectors/ts/src/avro.ts
// Minimal Apache Avro binary codec (the subset Ring needs), pure TypeScript.
// Encoding matches the Avro spec, so it interoperates with the Rust broker and
// the other connectors.

const U64 = (1n << 64n) - 1n;

export class Encoder {
  private bytes: number[] = [];

  private varint(zz: bigint): void {
    while (zz & ~0x7fn) {
      this.bytes.push(Number((zz & 0x7fn) | 0x80n));
      zz >>= 7n;
    }
    this.bytes.push(Number(zz & 0x7fn));
  }

  putLong(v: number | bigint): this {
    const b = BigInt(v);
    this.varint(((b << 1n) ^ (b >> 63n)) & U64);
    return this;
  }
  putInt(v: number): this {
    return this.putLong(v);
  }
  putBool(v: boolean): this {
    this.bytes.push(v ? 1 : 0);
    return this;
  }
  putDouble(v: number): this {
    const buf = new ArrayBuffer(8);
    new DataView(buf).setFloat64(0, v, true);
    for (const x of new Uint8Array(buf)) this.bytes.push(x);
    return this;
  }
  putBytes(b: Uint8Array): this {
    this.putLong(b.length);
    for (const x of b) this.bytes.push(x);
    return this;
  }
  putString(s: string): this {
    return this.putBytes(new TextEncoder().encode(s));
  }
  arrayStart(count: number | bigint): this {
    return this.putLong(count);
  }
  arrayEnd(): this {
    return this.putLong(0);
  }
  bytes_(): Uint8Array {
    return new Uint8Array(this.bytes);
  }
}

export class Decoder {
  private pos = 0;
  constructor(private data: Uint8Array) {}

  private varint(): bigint {
    let result = 0n;
    let shift = 0n;
    for (;;) {
      const b = this.data[this.pos++];
      result |= BigInt(b & 0x7f) << shift;
      if (!(b & 0x80)) break;
      shift += 7n;
    }
    return result;
  }

  long(): bigint {
    const n = this.varint();
    return (n >> 1n) ^ -(n & 1n);
  }
  int(): number {
    return Number(this.long());
  }
  bool(): boolean {
    return this.data[this.pos++] !== 0;
  }
  double(): number {
    const dv = new DataView(this.data.buffer, this.data.byteOffset + this.pos, 8);
    this.pos += 8;
    return dv.getFloat64(0, true);
  }
  bytes(): Uint8Array {
    const n = Number(this.long());
    const out = this.data.slice(this.pos, this.pos + n);
    this.pos += n;
    return out;
  }
  string(): string {
    return new TextDecoder().decode(this.bytes());
  }
  arrayCount(): bigint {
    return this.long();
  }
}

File: connectors/go/avro.go
package impulsering

import (
	"encoding/binary"
	"math"
)

// Encoder builds an Apache Avro binary datum (the subset Ring needs).
type Encoder struct{ Buf []byte }

func NewEncoder() *Encoder { return &Encoder{} }

func (e *Encoder) putVarint(zz uint64) {
	for zz&^uint64(0x7f) != 0 {
		e.Buf = append(e.Buf, byte((zz&0x7f)|0x80))
		zz >>= 7
	}
	e.Buf = append(e.Buf, byte(zz))
}

// PutLong writes a zig-zag varint long (Avro int and long share this encoding).
func (e *Encoder) PutLong(v int64) { e.putVarint(uint64((v << 1) ^ (v >> 63))) }

func (e *Encoder) PutInt(v int32) { e.PutLong(int64(v)) }

func (e *Encoder) PutBool(v bool) {
	if v {
		e.Buf = append(e.Buf, 1)
	} else {
		e.Buf = append(e.Buf, 0)
	}
}

func (e *Encoder) PutDouble(v float64) {
	var b [8]byte
	binary.LittleEndian.PutUint64(b[:], math.Float64bits(v))
	e.Buf = append(e.Buf, b[:]...)
}

func (e *Encoder) PutBytes(b []byte) {
	e.PutLong(int64(len(b)))
	e.Buf = append(e.Buf, b...)
}

func (e *Encoder) PutString(s string) { e.PutBytes([]byte(s)) }

func (e *Encoder) Bytes() []byte { return e.Buf }

// Decoder reads an Avro binary datum.
type Decoder struct {
	data []byte
	pos  int
}

func NewDecoder(data []byte) *Decoder { return &Decoder{data: data} }

func (d *Decoder) varint() uint64 {
	var result uint64
	var shift uint
	for d.pos < len(d.data) {
		b := d.data[d.pos]
		d.pos++
		result |= uint64(b&0x7f) << shift
		if b&0x80 == 0 {
			break
		}
		shift += 7
	}
	return result
}

func (d *Decoder) Long() int64 {
	n := d.varint()
	return int64(n>>1) ^ -int64(n&1)
}

func (d *Decoder) Int() int32 { return int32(d.Long()) }

func (d *Decoder) Bool() bool {
	if d.pos >= len(d.data) {
		return false
	}
	b := d.data[d.pos]
	d.pos++
	return b != 0
}

func (d *Decoder) Double() float64 {
	if d.pos+8 > len(d.data) {
		return 0
	}
	v := math.Float64frombits(binary.LittleEndian.Uint64(d.data[d.pos:]))
	d.pos += 8
	return v
}

func (d *Decoder) Bytes() []byte {
	n := int(d.Long())
	if n < 0 || d.pos+n > len(d.data) {
		n = 0
	}
	out := make([]byte, n)
	copy(out, d.data[d.pos:d.pos+n])
	d.pos += n
	return out
}

func (d *Decoder) String() string { return string(d.Bytes()) }

// ArrayCount reads a block count; 0 ends the array.
func (d *Decoder) ArrayCount() int64 { return d.Long() }

// peekLong reads the first Avro long of a body without consuming a Decoder.
func peekLong(body []byte) int64 { return NewDecoder(body).Long() }
